lib/include/jxl/decode.h

```diff
@@ -913,6 +950,142 @@ JXL_EXPORT JxlDecoderStatus JxlDecoderSetJPEGBuffer(JxlDecoder* dec,
+ *
+ * Calling JxlDecoderReleaseBoxBuffer is required whenever
+ * a buffer is already set and a new buffer needs to be added with
+ * JxlDecoderSetBoxBuffer, but is not required before JxlDecoderDestroy or
+ * JxlDecoderReset.
+ *
+ * Calling JxlDecoderReleaseBoxBuffer when no buffer is set is
+ * not an error and returns 0.
+ *
+ * @param dec decoder object
+ * @return the amount of bytes the decoder has not yet written to of the data
+ * set by JxlDecoderSetBoxBuffer, or 0 if no buffer is set or
+ * JxlDecoderReleaseBoxBuffer was already called.
+ */
+JXL_EXPORT size_t JxlDecoderReleaseBoxBuffer(JxlDecoder* dec);
+
+/**
+ * Outputs the type of the current box, after a JXL_DEC_BOX event occured, as 4
+ * characters plus a null termination character. If the box is of type "brob",
+ * the underlying uncompressed box type is returned instead, not "brob".
+ *
+ * @param dec decoder object
+ * @param type buffer to copy the type into
+ * @return JXL_DEC_SUCCESS if the value is available, JXL_DEC_ERROR if not, for
+ *    example the JXL file does not use the container format.
+ */
+JXL_EXPORT JxlDecoderStatus JxlDecoderGetBoxType(JxlDecoder* dec, char type[5]);
+
+/**
+ * Outputs whether the given type equals the type returned by
+ * JxlDecoderGetBoxType.
+ *
+ * @param dec decoder object
+ * @param type box type to compare with
+ * @return JXL_TRUE if the box type is available and equal to the given type,
+ * JXL_FALSE otherwise.
+ */
+JXL_EXPORT JXL_BOOL JxlDecoderBoxTypeEquals(JxlDecoder* dec, const char* type);
+
+/**
+ * Outputs the type of the current box, after a JXL_DEC_BOX event occured, as 4
+ * characters plus a null termination character. If the box is of type "brob",
+ * the underlying uncompressed box type is returned instead, not "brob".
+ *
+ * @param dec decoder object
+ * @param type buffer to copy the type into
+ * @return JXL_DEC_SUCCESS if the value is available, JXL_DEC_ERROR if not, for
+ *    example the JXL file does not use the container format.
+ */
+JXL_EXPORT JxlDecoderStatus JxlDecoderGetBoxTypeRaw(JxlDecoder* dec,
+                                                    char type[5]);
+
+/**
+ * Outputs whether the given type equals the type returned by
+ * JxlDecoderGetBoxTypeRaw.
+ *
+ * @param dec decoder object
+ * @param type box type to compare with
+ * @return JXL_TRUE if the raw box type is available and equal to the given
+ * type, JXL_FALSE otherwise.
+ */
+JXL_EXPORT JXL_BOOL JxlDecoderBoxTypeRawEquals(JxlDecoder* dec,
+                                               const char* type);
+
+/**
+ * Outputs whether the raw box type equals "brob", indicating it is a brotli
+ * compressed box. Its underlying type is still returned with
+ * JxlDecoderGetBoxType. The data returned to the box buffer is transparently
+ * decompressed, unless JxlDecoderSetBoxModeRaw is enabled, so this function is
+ * not needed to just get the data, as long as the library is compiled with
+ * brotli.
```

Comment:
I'd consider moving all the "brotli-only" functions in a separate part of the header, and writing comments of the form "The following functions are only available if JXL_SMALL_DECODER_LIBRARY is not set" (or whatever), as well as putting them behind an `#if`.

Reply:
We could make them unavailable at all at compile time, returning a compilation error, or instead we could make them exist in the API, but return "error" or "false" if brotli is not compiled in.

Not available behind #if has the advantage that you detect the problem sooner, but the disadvantage that it may break compilation of external libraries, even if they're only using the brotli decompression for non-core reasons.

Any thought on this?

Reply:
Disabling brotli is not meant for compiled libraries that are to be distributed but rather for squeezing as many bytes as we can out of a static build - I wouldn't be concerned about external libraries.

Reply:
It now only requires it for decompressed mode. Having the small vs non-small library with brotli dependency is for a different pull request.